### ISSUE TITLE
[Feature] Analyse segments of files (to aid CUE support)

### ISSUE THREAD

**Issue by CDrummond** [Polochon-street/bliss-rs#34]
I'm using bliss-rs analysis to provide a mix of similar tracks for LMS. LMS, as well as MPD, support reading tracks from CUE files, but bliss-rs does not. In my analysis code ([https://github.com/CDrummond/bliss-analyser/blob/master/src/analyse.rs](url)) I use ffmpeg (via calling ffmpeg executable) to create temporary files for each track within a CUE's main file. It would be great, however, if bliss-rs supported this itself.

The simplest way would be somehing like (forgive the syntax, I'm very new to Rust):

```
pub fn Song::from_segment(path: Path, start: Duration, length: Duration) -> BlissResult<Self> {
  ...
}
```

**Comment by Polochon-street** [comment]
Hi!
Indeed, cue support has been something that's been asked for for quite some time now, so I'll address it soon. Instead of the signature you pasted though, how would something like this

`pub fn from_cue(path: PathBuf,  Vec<Track>) -> Vec<BlissResult<Self>>` work? With `Track` be from the `cue` library.

Then we could just extract all Songs in one decoding instead of having to run a bunch of seek calls.
I don't really use CUE sheets, so I'd really be interested in your feedback before starting implementing it :)

**Comment by CDrummond** [comment]
Must admit I do not use CUEs myself either. I had 4 CUE albums, but decided it was just easier to manually split them up :) When I released my bliss based mixer for LMS a user asked about CUE file support - so I decided to try to implement this.

Why not just `pub fn from_cue(path: PathBuf, Vec<Duration>) -> Vec<BlissResult<Self>>` i.e. just pass a list of start times?

**Comment by Polochon-street** [comment]
hmm, since this kind of function will be specific to CUE files, why not just `pub fn from_cue(path_to_cue_file: PathBuf) -> Vec<BlissResult<Self>>`?

Then I could just use this https://docs.rs/cue/latest/cue/cd/struct.CD.html to return `Song`s with the right info for title, artist, album, etc.
What do you think?

**Comment by CDrummond** [comment]
Ah, oops! I forgot `Song` already had meta data :( Any chance you could add `album artist` and `duration` ? I think these being missing is the reason I use lofty-rs to read tags. If not I can continue to use lofty-rs, but it does mead I'd also need to parse the CUE file to get `album artist` and durations.

My mixer uses `album artist` to allow tracks from the same album to not be used for X tracks. Likewise I use duration so that too short or too long tracks can be excluded.

**Comment by Polochon-street** [comment]
It shouldn't be too difficult to add :)
Could you open another issue so it's easier for me to track these?

**Comment by Polochon-street** [comment]
I've made a `Song::from_cue` function; would you be able to hop on the `add-cue-support` branch and try it out? (https://github.com/Polochon-street/bliss-rs/pull/38)

Just want to make sure it's usable before merging it - I've written tests using a proper CUE file / corresponding WAV file, but maybe you'll find shortcomings using it in the real world :)

**Comment by Polochon-street** [comment]
For `analyze_paths_streaming` - I've moved it over to `lib.rs`, as `analyze_paths`, since it looks more standard. I think I've written a little something on the CHANGELOG for that :grin: 

For `rcue`, I wanted to use it, but they have no crates released, which means that I can't both use it and release new bliss versions to `crates.io`. I've opened an issue here https://github.com/gyng/rcue/issues/9, but until then, there's not much I can do.
It's too bad, because the library looks *really* neat.

**Comment by CDrummond** [comment]
Ah, sorry I should have looked closer :(

Shame about rcue, but I really don't want to add libcue as a dependency for my analyser. I guess seeing as this feature is optional I can just not support it for now - even though I was the one asking for it.

**Comment by Polochon-street** [comment]
I'll just wait til `rcue` gets a release - hopefully we'll have an answer in the upcoming days :crossed_fingers:

**Comment by CDrummond** [comment]
Yeah, sorry for being a pain! I do appreciate the work you have put into implementing this. And to bliss-rs itself, I use this almost every day.

**Comment by Polochon-street** [comment]
You're not being a pain at all, I needed to implement CUE support at some point anyways, this feature's been repeatedly asked for :)

As a consolation prize, I've added album_artist and duration to the Song's fields :grin: everything should work, but ping me if there's a problem somehow

**Comment by CDrummond** [comment]
...also, will `analyze_paths` support passing in CUE files? That way only 1 call is required :)

**Comment by Polochon-street** [comment]
yeah, it seems pretty straightforward to make it so :)

**Comment by Polochon-street** [comment]
hey again, the `rcue` crate has been released :tada:
I've made something with it for CUE support on this branch: https://github.com/Polochon-street/bliss-rs/pull/40

About implementing it directly in `analyze_paths`, though, just want to make sure: you'd always give only the CUE file, and not the actual FLAC / WAV file in the file list, right? I'm afraid of duplicates

I know MPD handles the CUE by listing tracks like `path/to/file/actual_cue_file.cue/track0001`, etc, so my MPD plugin will have to fiddle a bit with the list to make it just submitting `path/to/file/actual_cue_file.cue`. Will it be the same for your plug-in?

**Comment by CDrummond** [comment]
Brilliant, thanks! Will test later, and integrate into my code, when I get the chance.

Yes, I'd always just pass the CUE file.

My plugin is actually 3 parts; bliss-analyser, bliss-mixer, and lms-blissmixer. bliss-analyser is a Rust app that uses your library to get analysis results. This creates an SQLite database with track path, metadata, and analysis results. For CUE file tracks I save as `<cue audio file>.CUE_TRACK.<track number>` (e.g. `Some Artist/Some Album/Album.flac.CUE_TRACK.1`). bliss-mixer is a Rust app that reads the SQLite DB to provided mixes based upon seed tracks. lms-blissmixer is a Perl plugin (as LMS is perl) that starts bliss-mixer and sends seed tracks to get the mixes.

The reason that `<cue audio file>.CUE_TRACK.<track number>` is uses is that LMS stores CUE tracks as (e.g.) `file://path/Some%20Artist/Some%20Album/Album.flac#0-182.0`

**Comment by Polochon-street** [comment]
thanks for the explanation :) out of curiosity, if LMS stores CUE tracks as `Album.flac#0-182.0`, how do you know the path of the actual CUE file?

Just to make sure we're on the same page - if you pass e.g. `["track1.flac", "track2.flac", "cue_file.cue"]`, then I would output `[Song { path: "track1.flac" }, Song { path: "track2.flac" }, Song { path: "cue_file.flac/CUE_TRACK001" }, Song { path: "cue_file.flac/CUE_TRACK002" }, ...]`
How does that look? Will you be able to make the correspondence between "cue_file.flac/CUE_TRACK001" and "cue_file.flac#01"?

**Comment by CDrummond** [comment]
My plugin will have the LMS track object, this has the full path to the audio file and the track number. Hence I'll map to `blah/blah/blah.flac.CUE_TRACK.1`

If you are going to use `Song { path: "cue_file.flac/CUE_TRACK002" }` then that makes things easier for me. I thought it would have been `Song { path: "cue_file.cue/CUE_TRACK002" }` (i.e. cue file not audio file) - which would have meant I would need to store a map from audio-file to cue-file.

I would change my code to use `/CUE_TRACK<00 num>` - but I fear `.CUE_TRACK.<num>` is already been used. Still, conversion will be trivial.

**Comment by Polochon-street** [comment]
okay, I think I know what I'll do. I'll add a `pub cue_info: Option<CueInfo>` field on each `Song` that will contain: 1. the cue file location 2. the audio file location 3. extra info?

That way you can just format the path the way you want since you'll have access to all this info, without having to parse it from the path string. How does that sound?

**Comment by CDrummond** [comment]
Sounds great to me :)

**Comment by Polochon-street** [comment]
alright - I've updated https://github.com/Polochon-street/bliss-rs/pull/40 so that `analyze_paths` also analyses songs from passed CUE files. It's not perfect, but I think that would be nice if you could try and see if it works for you :)

**Comment by CDrummond** [comment]
Can't access the fields of CueInfo:

```
field `audio_file_path` of struct `bliss_audio::song::CueInfo` is private
```

**Comment by CDrummond** [comment]
I've tested the new branch, and it seems to work. However, the issue above needs to be resolved (I used the CUE path for me tests).

**Comment by Polochon-street** [comment]
Hi,
I've just pushed the fix. Tell me whether it works for you; if so, I'll make a new bliss release that will include this shortly.

**Comment by Polochon-street** [comment]
Alright, I've cleaned this up, and released a new version. Close this issue if everything is :100: for you :)

**Comment by CDrummond** [comment]
Works for me!
